source/der.c

```diff
@@ -0,0 +1,452 @@
+        if (!aws_byte_buf_write_be16(buf, (uint16_t)len)) {
+            return AWS_OP_ERR;
+        }
+    } else if (len > INT8_MAX) {
+        /* Write the high bit + 1 byte length */
+        if (!aws_byte_buf_write_u8(buf, 0x81)) {
+            return AWS_OP_ERR;
+        }
+        if (!aws_byte_buf_write_u8(buf, (uint8_t)len)) {
+            return AWS_OP_ERR;
+        }
+    } else {
+        if (!aws_byte_buf_write_u8(buf, (uint8_t)len)) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    switch (tlv->tag) {
+        case AWS_DER_INTEGER: {
+            /* if the first byte has the sign bit set, insert an extra 0x00 byte to indicate unsigned */
+            uint8_t first_byte = tlv->value[0];
+            if (first_byte & 0x80) {
+                if (!aws_byte_buf_write_u8(buf, 0)) {
+                    return AWS_OP_ERR;
+                }
+            }
+            if (!aws_byte_buf_write(buf, tlv->value, tlv->length)) {
+                return AWS_OP_ERR;
+            }
+        } break;
+        case AWS_DER_BOOLEAN:
+            if (!aws_byte_buf_write_u8(buf, (*tlv->value) ? 0xff : 0x00)) {
+                return AWS_OP_ERR;
+            }
+            break;
+        case AWS_DER_BIT_STRING:
+            /* Write that there are 0 skipped bits */
+            if (!aws_byte_buf_write_u8(buf, 0)) {
+                return AWS_OP_ERR;
+            }
+            /* FALLTHROUGH */
+        case AWS_DER_BMPString:
+        case AWS_DER_IA5String:
+        case AWS_DER_PrintableString:
+        case AWS_DER_UTF8_STRING:
+        case AWS_DER_OBJECT_IDENTIFIER:
+        case AWS_DER_OCTET_STRING:
+        case AWS_DER_SEQUENCE:
+        case AWS_DER_SET:
+            if (!aws_byte_buf_write(buf, tlv->value, tlv->length)) {
+                return AWS_OP_ERR;
+            }
+            break;
+        case AWS_DER_NULL:
+            /* No value bytes */
+            break;
+        default:
+            AWS_FATAL_ASSERT(!"TLV tag is not a supported encoding type");
+            return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_der_encoder_init(struct aws_der_encoder *encoder, struct aws_allocator *allocator, size_t capacity) {
+    encoder->allocator = allocator;
+    if (aws_byte_buf_init(&encoder->storage, encoder->allocator, capacity)) {
+        return AWS_OP_ERR;
+    }
+    if (aws_array_list_init_dynamic(&encoder->stack, encoder->allocator, 4, sizeof(struct der_tlv))) {
+        return AWS_OP_ERR;
+    }
+
+    encoder->buffer = &encoder->storage;
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_der_encoder_clean_up(struct aws_der_encoder *encoder) {
+    aws_byte_buf_clean_up_secure(&encoder->storage);
+    aws_array_list_clean_up(&encoder->stack);
+}
+
+int aws_der_encoder_write_integer(struct aws_der_encoder *encoder, struct aws_byte_cursor integer) {
+    AWS_FATAL_ASSERT(integer.len <= UINT32_MAX);
+    struct der_tlv tlv = {
+        .tag = AWS_DER_INTEGER,
+        .length = (uint32_t)integer.len,
```

Comment:
easier to just make der_tlv.length be a size_t?
rather than need to check that cursor.len doesn't exceed UINT32_MAX in a bunch of places?

Reply:
The format is 32 bit lengths, so we have to check it anyway